### ISSUE TITLE
Curl crash on iOS with DoH and http/2 enabled

### ISSUE THREAD

**Issue by lagner** [curl/curl#8192]
**Library version**
_curl_ 7.79.1 
_libnghttp2_ 1.45.1
_c-ares_ 1.17.2

The combination of these libraries works stable. But we have some issues with name resolving on Android and iOS. Therefore we try to use the DoH.
I added the following lines
```cpp
namespace {

const auto PREDEFINED_HOSTS = []() -> auto {
	curl_slist * hosts = nullptr;
	curl_slist_append(hosts, "cloudflare-dns.com:443:1.1.1.1,1.0.0.1");
	curl_slist_append(hosts, "dns.google:443:8.8.8.8,8.8.4.4");
	return CurlListPtr(hosts, CurlListDeleter());
}();

} // namespace

void send_request() {
    // ....
    curl_easy_setopt(curl, CURLOPT_RESOLVE, PREDEFINED_HOSTS.get());
    curl_easy_setopt(curl, CURLOPT_DOH_URL, "https://cloudflare-dns.com/dns-query");
    curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1);

    // ....
}
```
It's C++ not C. Sorry about that but the idea is clear I hope.

Since that the application has started crashing on iOS randomly.
The callstack is the same for all crashes
```
#0	0x000000010a1d586c in on_begin_headers at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/http2.c:935
#1	0x000000010a231690 in session_call_on_begin_headers at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libnghttp2/1.45.1/sdk/release/build/faf309cc256628e27ef010c9360c9c161339ff1f/source_subfolder/lib/nghttp2_session.c:3325
#2	0x000000010a2317b8 in nghttp2_session_on_response_headers_received at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libnghttp2/1.45.1/sdk/release/build/faf309cc256628e27ef010c9360c9c161339ff1f/source_subfolder/lib/nghttp2_session.c:3958
#3	0x000000010a236420 in session_process_headers_frame at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libnghttp2/1.45.1/sdk/release/build/faf309cc256628e27ef010c9360c9c161339ff1f/source_subfolder/lib/nghttp2_session.c:4082
#4	0x000000010a233eac in nghttp2_session_mem_recv at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libnghttp2/1.45.1/sdk/release/build/faf309cc256628e27ef010c9360c9c161339ff1f/source_subfolder/lib/nghttp2_session.c:5589
#5	0x000000010a1d2d00 in h2_process_pending_input at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/http2.c:1378
#6	0x000000010a1d4ef4 in http2_connisdead at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/http2.c:203
#7	0x000000010a1d4c78 in http2_conncheck at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/http2.c:244
#8	0x000000010a214cd4 in extract_if_dead at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/url.c:1013
#9	0x000000010a214c1c in call_extract_if_dead at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/url.c:1046
#10	0x000000010a1a8f00 in Curl_conncache_foreach at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/conncache.c:354
#11	0x000000010a212354 in prune_dead_connections at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/url.c:1075
#12	0x000000010a20fb94 in create_conn at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/url.c:3839
#13	0x000000010a20f2f4 in Curl_connect at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/url.c:4099
#14	0x000000010a1e9258 in multi_runsingle at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/multi.c:1814
#15	0x000000010a1e8f80 in curl_multi_perform at /Users/s.lagner/Projects/conan-ci/build/conan/.conan/data/libcurl/7.79.1/sdk/release/build/922d474998d0224f6934f47c3ae8b7f43de9171e/source_subfolder/lib/multi.c:2568
```
I'm trying to build a small example to reproduce. But maybe it's a known issue and somebody knows how to work around it.

**Comment by jay** [comment]
[curl_slist_append](https://curl.se/libcurl/c/curl_slist_append.html) you should check the return value. Further if list is null and you call curl_slist_append(list, "foo"); then you must assign it otherwise list stays null as it does in your code. also i notice you're returning an object, CurlListPtr, I don't know what that is but I guess it has a destructor, and you're passing that to [CURLOPT_RESOLVE](https://curl.se/libcurl/c/CURLOPT_RESOLVE.html) which takes `struct curl_slist *` only.

The proper use of both of those is as described in the example sections of those docs. And I don't know how http2 fits in here but maybe the way you are handling lists is somehow causing that

**Comment by bagder** [comment]
If you still suffer from a problem after you've fixed your code, it would help if you would provide a complete source code that reproduces the issue. This is not a known issue. Without more details I don't think there's a lot we can do.

**Comment by lagner** [comment]
Thank you for replying
> curl_slist_append you should check the return value. Further if list is null and you call curl_slist_append(list, "foo"); then you must assign it otherwise list stays null as it does in your code

You are right. I've lost an assignment during refactoring. 
One more try
```
const auto PREDEFINED_HOSTS = []() -> auto {
	curl_slist * hosts = curl_slist_append(nullptr, "cloudflare-dns.com:443:1.1.1.1,1.0.0.1");
	UNI_ASSERT(hosts); // sends SIGABORT if hosts is NULL
	return CurlListPtr(hosts, CurlListDeleter());
}();
```
Unfortunately, it does not help. The crash is still here.

> then you must assign it otherwise list stays null as it does in your code. also i notice you're returning an object, CurlListPtr, I don't know what that is but I guess it has a destructor, and you're passing that to CURLOPT_RESOLVE which takes struct curl_slist * only.

I tried to simplify the example. 
The following code will clarify, I hope
```cpp
struct CurlListDeleter {
	void operator()(curl_slist * ptr) const {
		curl_slist_free_all(ptr);
	}
};

using CurlListPtr = std::unique_ptr<curl_slist, CurlListDeleter>;
```
So `PREDEFINED_HOSTS.get()` is `curl_slist *` exactly.
The screenshot is just in case
<img width="1503" alt="Screenshot 2021-12-28 at 12 15 17" src="https://user-images.githubusercontent.com/1218917/147536456-78e7e77c-b00a-484d-9b44-429bd7733c1a.png">

I'm still debugging. Poor knowledge of curl internals prevents me from doing effectively

**Comment by bagder** [comment]
The `data->req.p.http` pointer you get there looks like it is already freed. I would check what stream that is and how come that pointer is bad! `stream_s` seems like to point to a proper struct...

**Comment by jay** [comment]
also please review [libcurl thread safety guidelines](https://curl.se/libcurl/c/threadsafe.html)

**Comment by lagner** [comment]
Sorry for the late reply, long holidays. 

@jay thank you for the link. I've read it carefully and I think we do it right.
A separate thread runs only for the Curl purposes. Nobody out of this thread can touch the Curl API. In addition, we set
```
curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1);
```
  

_One related question_. 
Our application uses threads heavily. That's why I am sure we have to set CURLOPT_NOSIGNAL=1. And now we're going to use DoH for all queries.
Taking it into account, can I exclude the c-ares from the build safely?
In other words, is DoH with CURLOPT_NOSIGNAL=1 enough or do I still need an asynchronous DNS resolver?

**Comment by jay** [comment]
@lagner we will need a minimal self contained reproducible example to pursue this further.

> Our application uses threads heavily. That's why I am sure we have to set CURLOPT_NOSIGNAL=1. And now we're going to use DoH for all queries.
> Taking it into account, can I exclude the c-ares from the build safely?

curl needs to be built with a default resolver that is not DoH. The resolver can be asynchronous (c-ares or threaded) or synchronous. Which resolver you build with should not make any difference whether DoH (DNS-over-HTTPS) works. The DoH server name is resolved using the default resolver and other resolves are done via DoH. You can force the DoH server name to be resolved using [CURLOPT_RESOLVE](https://curl.se/libcurl/c/CURLOPT_RESOLVE.html), if you are willing to take the risk that the ip addresses won't change for example dns.google:443:8.8.8.8,8.8.4.4
